appmgr/src/setup.rs

```diff
@@ -222,12 +228,27 @@ pub async fn complete(#[context] ctx: SetupContext) -> Result<(), Error> {
         ));
     };
     if tokio::fs::metadata(PRODUCT_KEY_PATH).await.is_err() {
-        let mut pkey_file = File::create(PRODUCT_KEY_PATH).await?;
-        pkey_file
-            .write_all(ctx.product_key().await?.as_bytes())
-            .await?;
-        pkey_file.sync_all().await?;
+        crate::hostname::set_product_key(&*ctx.product_key().await?).await?;
+    } else {
+        let key_on_disk = crate::hostname::get_product_key().await?;
+        let key_in_cache = ctx.product_key().await?;
+        if *key_in_cache != key_on_disk {
+            crate::hostname::set_product_key(&*ctx.product_key().await?).await?;
+        }
```

Comment:
why would these ever differ?

Reply:
in a case where someone built from source without NO_KEY and then they tried to migrate from 0.2.x

Reply:
is this a case we are trying to support?

Reply:
It is imperative that we don't just blackhole people's product keys that they have purchased from us. We need to make sure that the "correct" product key is in place when migrating. If they also built 02x from source and generated their own key then, this process does nothing to harm that case, but it does fix the cases where a user migrates from an old system and they built 0.3 from source. So yes. I think it is something we are trying to support.